[ERM UI] Add pagination to linked/downstream projects tab

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SS-103

This is the first PR in a series of PRs that will make subtle changes/additions to the Linked Projects UI. This PR accomplishes the following: 
- adds pagination to the downstream/linked projects table
- adds confirmation modal when removing domain links

![Screen Shot 2021-05-31 at 11 11 47 AM](https://user-images.githubusercontent.com/15785053/120213410-28312f80-c201-11eb-9373-1323952bbe20.png)
![Screen Shot 2021-05-31 at 11 12 14 AM](https://user-images.githubusercontent.com/15785053/120213416-2bc4b680-c201-11eb-97fc-f45ddbf72df4.png)


## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`LINKED_DOMAINS`
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
UI changes.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Will run all UI changes through QA when complete.
### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Locally tested.
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
add pagination to domain links table

gherceg committed:
rename Delete to Remove

gherceg committed:
fix total items bug

gherceg committed:
use modals correctly

gherceg committed:
ensure all text is marked with trans

orangejenny commented on corehq/apps/linked_domain/templates/linked_domain/domain_links.html:
```diff
@@ -102,10 +102,10 @@ <h2>{% trans "Project Link" %}</h2>  {# header for the only tab that will be dis
                       <div class="modal-content">
                         <div class="modal-header">
                           <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
```
ARIA labels can also be translated, they get read aloud by screen readers.

orangejenny commented on corehq/apps/linked_domain/templates/linked_domain/domain_links.html:
```diff
@@ -88,15 +88,32 @@ <h2>{% trans "Project Link" %}</h2>  {# header for the only tab that will be dis
+                        </div>
+                        <div class="modal-footer">
+                          <button type="button" class="btn btn-default" data-dismiss="modal">{% trans 'Cancel' %}</button>
+                          <button type="button" class="btn btn-primary" data-bind="click: $root.deleteLink.bind($data)" data-dismiss="modal">{% trans 'Remove' %}</button>
```
`btn-danger` instead of `btn-primary`, becuase this is a destructive action

orangejenny commented on corehq/apps/linked_domain/static/linked_domain/js/domain_links.js:
```diff
@@ -66,33 +67,50 @@ hqDefine("linked_domain/js/domain_links", [
             }
         }
 
+        self.domain_links = ko.observableArray(_.map(data.linked_domains, function (link) {
```
Nitpick: you can just do `_.map(data.linked_domains, DomainLink)`

orangejenny commented on corehq/apps/linked_domain/static/linked_domain/js/domain_links.js:
```diff
@@ -66,33 +67,50 @@ hqDefine("linked_domain/js/domain_links", [
 
-        self.model_status = _.map(data.model_status, ModelStatus);
+        // manage downstream domains tab
+        self.paginated_domain_links = ko.observableArray([]);
```
This file is inconsistent, but the js naming convention is camel case for attributes.

gherceg replied:
Realized I never replied. I've gone back and fourth on updating existing variables but it results in enough changes to warrant its own PR. This is a good reminder to follow the js naming convention for any variables that I add though so thanks.

orangejenny commented on corehq/apps/linked_domain/static/linked_domain/js/domain_links.js:
```diff
@@ -139,6 +157,14 @@ hqDefine("linked_domain/js/domain_links", [
         setRMI(initialPageData.reverse('linked_domain:domain_link_rmi'), csrfToken);
 
         var model = DomainLinksViewModel(view_data);
-        $("#domain_links").koApplyBindings(model);
```
Curious why this changed, aren't all three tabs still inside the `#domain_links` container?

gherceg replied:
Mix of reasons. I initially started this change using a crud paginated view, and the best way to include that forced me to make this change to avoid applying bindings to the same element twice. Obviously did not continue with that approach, but the idea of separating each tab into its own knockout element seemed ideal. We currently pass in one big view model to #domain_links like you said, but a shift to smaller, more isolated view models would be nice. 

Given that in this PR there aren't any benefits from making this change, I can understand your question. Would love to hear your thoughts if you are opposed to trying to move this direction.

orangejenny replied:
Got it. Smaller models sounds good, the big models do get unwieldy, and this page does seem like a good candidate for splitting up.

orangejenny commented on corehq/apps/linked_domain/templates/linked_domain/domain_links.html:
```diff
@@ -23,144 +23,156 @@
-                </tbody>
-              </table>
+        <div class="tab-pane fade in active" id="tabs-pull-content">
+          <div id="ko-tabs-pull-content" class="ko-template">
```
Nitpick: what's the extra layer for?

gherceg replied:
@biyeun might be able to speak better to this than I can, but I'll give it a shot. The intention is to separate the knockout element from the tab/bootstrap element. At one point I did have an issue with the tab not rendering appropriately and this change appeared to that fix that, but I'm not positive the same applies for the code in its current state.

orangejenny replied:
Interesting. I can see the logic of separating the styling and the interactivity.

orangejenny commented on corehq/apps/linked_domain/templates/linked_domain/domain_links.html:
```diff
@@ -23,144 +23,156 @@
+              params="goToPage: goToPage,
+                      perPage: itemsPerPage,
+                      totalItems: totalItems,
+                      slug: 'linked-domains-table',
```
I think this slug is just for a cookie if you're using one, which I don't think you are.

orangejenny reviewed (approved):
Left some nitpicky comments, but overall this looks good.

gherceg committed:
address feedback

- change btn-primary to btn-danger
- domain_links doesn't need to be observable
- folow js naming scheme for some vars
- translate aria label

gherceg committed:
appease stickler

gherceg committed:
address feedback correctly
